hs-bindgen/src/HsBindgen/C/AST/Literal.hs

```diff
@@ -0,0 +1,156 @@
+
+parseIntSuffix :: Parser IntSuffix
+parseIntSuffix = choice [
+      IntSuffixUnsigned <$ caseInsensitive "u"
+    , IntSuffixLongLong <$ caseInsensitive "ll"
+    , IntSuffixLong     <$ caseInsensitive "l"
+    , IntSuffixSize     <$ caseInsensitive "z"
+    ]
+
+parseIntLitAux :: Parser Integer
+parseIntLitAux = do
+    -- TODO: Should we keep the suffices around?
+    (base, digits, _suffices) <- aux
+
+    let multipliers :: [Integer]
+        multipliers = iterate (* baseToInt base) 1
+
+    return $ sum $ zipWith (*) (reverse $ mapMaybe getDigit digits) multipliers
+  where
+    aux :: Parser (Base, [Digit], [IntSuffix])
+    aux = do
+        base     <- parseBase
+        digits   <- many1 $ parseDigitInBase base
+        suffixes <- many parseIntSuffix
+        return (base, digits, suffixes)
+
+{-------------------------------------------------------------------------------
+  Auxiliary: integer representations in different bases
+-------------------------------------------------------------------------------}
+
+data Base =
+    BaseDec
+  | BaseOct
+  | BaseHex
+  | BaseBin
+  deriving stock (Show)
+
+baseToInt :: Base -> Integer
+baseToInt BaseDec = 10
+baseToInt BaseOct = 8
+baseToInt BaseBin = 2
+baseToInt BaseHex = 16
+
+-- | Digit in a given base
+data Digit = Digit Integer | Separator
+  deriving stock (Show)
+
+getDigit :: Digit -> Maybe Integer
+getDigit (Digit i) = Just i
+getDigit Separator = Nothing
+
+parseBase :: Parser Base
+parseBase = choice [
+      BaseHex <$ caseInsensitive "0x"
+    , BaseBin <$ caseInsensitive "0b"
+    , BaseOct <$ caseInsensitive "0"
+    , BaseDec <$ lookAhead (satisfy $ \c -> c >= '1' && c <= '9')
+    ]
+
+-- | Parse digit in the given base
+--
+-- Returns the value of the digit, of 'Nothing' for single quotes
+-- (which are allowed as a separator between digits).
+parseDigitInBase :: Base -> Parser Digit
+parseDigitInBase = satisfyWith . (. toLower) . aux
+  where
+    aux :: Base -> Char -> Maybe Digit
+    aux BaseDec c
+      | c >= '0' && c <= '9' = Just . Digit $ c `relativeTo` '0'
+    aux BaseBin c
+      | c >= '0' && c <= '1' = Just . Digit $ c `relativeTo` '0'
+    aux BaseOct c
+      | c >= '0' && c <= '7' = Just . Digit $ c `relativeTo` '0'
+    aux BaseHex c
+      | c >= '0' && c <= '9' = Just . Digit $ c `relativeTo` '0'
+      | c >= 'a' && c <= 'f' = Just . Digit $ c `relativeTo` 'a' + 10
+
+    aux _ '\'' = Just $ Separator
+    aux _ _    = Nothing
+
+    relativeTo :: Char -> Char -> Integer
+    relativeTo c r = fromIntegral $ ord c - ord r
+
+{-------------------------------------------------------------------------------
+  Auxiliary: parsec
+-------------------------------------------------------------------------------}
+
+type Parser = Parsec Text ()
+
+-- | Case-insensitive match
+--
+-- Does not consume any input unless the full string matches.
+caseInsensitive :: String -> Parser ()
+caseInsensitive = \str ->
+    try (mapM_ aux str <?> str)
```

Comment:
https://hackage.haskell.org/package/parsec-3.1.17.0/docs/Text-Parsec-Char.html#v:string-39-

Reply:
Yeah, but I wasn't sure how to avoid it here. I looked at `string'`, but it's implemented in terms of `tokens'`, which is not general enough (it uses equality on tokens, which is not what I want), and I can't copy and paste it to generalize it (I tried), because it uses features that aren't in the public API. So I wan't sure what to do; do you have a suggestion?

Reply:
Don't treat `parsec` as a context-free grammar. It's not. It's PEG.

```
0 -> (x -> hex | b -> bin | 1 .. 9 -> oct)
1 -> dec
```

Not as pretty as what you have now, but that is the price to pay.

Reply:
And for the future there is https://hackage.haskell.org/package/parsec-3.1.17.0/docs/Text-Parsec-Prim.html#v:mkPT which is close enough to using `ParsecT` constructor directly

Reply:
I will take a look at this `try` thing, but going to merge this PR to avoid blocking other work.

Reply:
Addressed in #197.